<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [3]:
import pandas as pd
import psycopg2

In [4]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

In [5]:
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [169]:
connection = psycopg2.connect(
    dbname = DBNAME,
    user = USER,
    password = PASSWORD,
    host = HOST,
    port = PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [7]:
# текст запроса
query_3_1 = f'''
SELECT COUNT(*) AS total_vacancies FROM vacancies
'''

In [8]:
# результат запроса
res = pd.read_sql_query(query_3_1, connection)
res

/var/folders/pz/qxrk46657v91nfg00jfysdx40000gn/T/ipykernel_13414/3716456928.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  res = pd.read_sql_query(query_3_1, connection)


total_vacancies
0            49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [9]:
# текст запроса
query_3_2 = f'''
SELECT COUNT(*) AS total_employers FROM employers
'''

In [10]:
# результат запроса
res = pd.read_sql_query(query_3_2, connection)
res

/var/folders/pz/qxrk46657v91nfg00jfysdx40000gn/T/ipykernel_13414/4117432493.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  res = pd.read_sql_query(query_3_2, connection)


total_employers
0            23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [11]:
# текст запроса
query_3_3 = f'''
SELECT COUNT(*) AS total_areas FROM areas
'''

In [12]:
# результат запроса
res = pd.read_sql_query(query_3_3, connection)
res

/var/folders/pz/qxrk46657v91nfg00jfysdx40000gn/T/ipykernel_13414/1598006713.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  res = pd.read_sql_query(query_3_3, connection)


total_areas
0         1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [13]:
# текст запроса
query_3_4 = f'''
SELECT COUNT(*) AS total_industries FROM industries
'''

In [14]:
# результат запроса
res = pd.read_sql_query(query_3_4, connection)
res

/var/folders/pz/qxrk46657v91nfg00jfysdx40000gn/T/ipykernel_13414/3942647788.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  res = pd.read_sql_query(query_3_4, connection)


total_industries
0               294

***

### Выводы
- В среднем, каждый работодатель опубликовал более 2х вакансий;

- Некоторые работодатели расположены в одних и тех же регионах, так как количество работодателей превышает количество регионов в 17+ раз;

- Многие работодатели оперируют в одних и тех же индустриях, возможно являясь конкурентами. Это может помочь нашим соискателям договариваться о более высоких зарплатах, так как несколько компаний из одной индустрии могут претендовать на одного специалиста. Мы тут даже можем осознанно рекомендовать специалиста 2+ компаниям из одной индустрии, учитывая, что эта индустрия интересна нашему специалисту.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [167]:
# текст запроса
# in the query, joining vacancies and areas by area id and group result by area name
query_4_1 = f'''
SELECT a.name, COUNT(b.id) AS vacancies_cnt
FROM areas a
JOIN vacancies b
ON a.id = b.area_id
GROUP BY 1
ORDER BY 2 DESC
'''

In [170]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

/var/folders/pz/qxrk46657v91nfg00jfysdx40000gn/T/ipykernel_13414/2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


name  vacancies_cnt
0             Москва           5333
1    Санкт-Петербург           2851
2              Минск           2112
3        Новосибирск           2006
4             Алматы           1892
..               ...            ...
764       Тарко-Сале              1
765    Новоаннинский              1
766            Бирск              1
767           Сасово              1
768             Анна              1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [17]:
# текст запроса
query_4_2 = f'''
SELECT COUNT(*) AS vacancies_with_salary
FROM vacancies
WHERE salary_from IS NOT NULL
OR salary_to IS NOT NULL
'''

In [18]:
# результат запроса
res = pd.read_sql_query(query_4_2, connection)
res

/var/folders/pz/qxrk46657v91nfg00jfysdx40000gn/T/ipykernel_13414/3481911129.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  res = pd.read_sql_query(query_4_2, connection)


vacancies_with_salary
0                  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [19]:
# текст запроса
query_4_3 = f'''
SELECT ROUND(AVG(salary_from)) AS avg_salary_from, ROUND(AVG(salary_to)) AS avg_salary_to
FROM vacancies
'''

In [20]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

/var/folders/pz/qxrk46657v91nfg00jfysdx40000gn/T/ipykernel_13414/2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


avg_salary_from  avg_salary_to
0          71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [21]:
# текст запроса
query_4_4 = f'''
SELECT schedule, employment, COUNT(id) AS vacancies_cnt
FROM vacancies
GROUP BY 1, 2
ORDER BY 3 DESC
'''

In [22]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

/var/folders/pz/qxrk46657v91nfg00jfysdx40000gn/T/ipykernel_13414/2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  vacancies_cnt
0        Полный день     Полная занятость          35367
1   Удаленная работа     Полная занятость           7802
2      Гибкий график     Полная занятость           1593
3   Удаленная работа  Частичная занятость           1312
4     Сменный график     Полная занятость            940
5        Полный день           Стажировка            569
6     Вахтовый метод     Полная занятость            367
7        Полный день  Частичная занятость            347
8      Гибкий график  Частичная занятость            312
9        Полный день     Проектная работа            141
10  Удаленная работа     Проектная работа            133
11     Гибкий график           Стажировка            116
12    Сменный график  Частичная занятость            101
13  Удаленная работа           Стажировка             64
14     Гибкий график     Проектная работа             18
15    Сменный график           Стажировка             12
16    Вахтовый метод     Проектная работа              2
17    Сменный график     Проектная работа              1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [23]:
# текст запроса
query_4_5 = f'''
SELECT experience, COUNT(id) AS vacancies_cnt
FROM vacancies
GROUP BY 1
ORDER BY 2
'''

In [24]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

/var/folders/pz/qxrk46657v91nfg00jfysdx40000gn/T/ipykernel_13414/2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  vacancies_cnt
0         Более 6 лет           1337
1           Нет опыта           7197
2       От 3 до 6 лет          14511
3  От 1 года до 3 лет          26152

***

### Выводы
- Большинство вакансий приходится на города-миллионники, что весьма логично, ведь там больше компаний. Вакансии из больших городов могут быть интересны соискателям с периферии, готовым к релокации;

- В нашей выборки присутствуют не только российские, но и, как минимум, белорусские и казахские работодатели. Казахские вакансии с релокаций могут пользваться особой популярностью у соискателей из России и Беларуси, готовых к релокации;

- Вилка зарплат - примерно 40К рублей, но пока что эта изолированная информация не дает нам никаких ценных инсайтов. Нужно смотреть в контексте позиций, локаций и требований;

- Большинство вакансий ориентированы на полную занятость и предлагают полный рабочий день (считаем удаленку полным рабочим днем из дома или кафе, ну и понимаем, насколько по-настоящему гибок гибкий график);

- Наконец, пв основном, работодатели на HH ищут специалистов с опытом работы от 1 до 3 лет. **Гипотезы:** 
    - Люди без опыта мало кому нужны (относительно большой риск)
    - Работодатели считают, что хорошие специалисты с опытом 3+ лет и без HH найдут себе работу. А если человек с опытом 3+ лет сидит на HH, то работодатели могут думать, что на него низкий спрос на рынке труда, и он(а) не является хорошим специалистом. По субъективному опыту, рекомендации от HR агенств воспринимаются более доверительно, нежели заявки на HH - поэтому мы можем напрямую связываться с работодателями, которые ищут DS специалистов с опытом 1-3 лет, но с вилкой зарплаты мидла, и предлагать им наших соискателей с бОльшим опытом работы. 

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [25]:
# текст запроса
# joining employers and vacancies to search for TOP-5 empoloyer names by the amount of vacancies posted
query_5_1 = f'''
SELECT a.name AS employer, COUNT(b.id) AS vacancies_cnt
FROM employers a
JOIN vacancies b
ON a.id = b.employer_id
GROUP BY 1
ORDER BY 2 DESC
LIMIT 5
'''

In [26]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

/var/folders/pz/qxrk46657v91nfg00jfysdx40000gn/T/ipykernel_13414/3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


employer  vacancies_cnt
0         Яндекс           1933
1     Ростелеком            491
2       Тинькофф            444
3           СБЕР            428
4  Газпром нефть            331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [27]:
# текст запроса
# LEFT joining areas with employers and vacancies to see for what areas vacancies count is NULL despite having employers registered there
# grouping the result by area name
# ordering the result by NULL vacancies and then by the number of employers in the area in the descending order
query_5_2 = f'''
SELECT a.name AS region, COUNT(b.id) AS employers_cnt, COUNT(c.id) AS vacancies_cnt
FROM areas a
LEFT JOIN employers b ON a.id = b.area
LEFT JOIN vacancies c ON a.id = c.area_id
GROUP BY 1
ORDER BY 3 NULLS FIRST, 2 DESC
'''

In [28]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

/var/folders/pz/qxrk46657v91nfg00jfysdx40000gn/T/ipykernel_13414/2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


region  employers_cnt  vacancies_cnt
0                 Россия            410              0
1              Казахстан            207              0
2     Московская область             75              0
3     Краснодарский край             19              0
4     Ростовская область             18              0
...                  ...            ...            ...
1357         Новосибирск        1149438        1149438
1358              Алматы        1364132        1364132
1359               Минск        2354880        2354880
1360     Санкт-Петербург        6320667        6320667
1361              Москва       31272712       31272712

[1362 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [29]:
# текст запроса
# joining employers and vacancies to count number of different regions where the employer posted its vacancies
query_5_3 = f'''
SELECT a.name AS employer, COUNT(DISTINCT(b.area_id)) AS regions_cnt
FROM employers a
JOIN vacancies b
ON a.id = b.employer_id
GROUP BY 1
ORDER BY 2 DESC
'''

In [30]:
# результат запроса
res = pd.read_sql_query(query_5_3, connection)
res

/var/folders/pz/qxrk46657v91nfg00jfysdx40000gn/T/ipykernel_13414/178708835.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  res = pd.read_sql_query(query_5_3, connection)


employer  regions_cnt
0                      Яндекс          181
1                  Ростелеком          152
2                  Спецремонт          116
3      Поляков Денис Иванович           88
4                   ООО ЕФИН            71
...                       ...          ...
14761                  UniSol            1
14762            UNISTORY LLC            1
14763                   UNIT6            1
14764     United Distribution            1
14765              UNITEL LLC            1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [31]:
# текст запроса
# first, creating a CTE for the table of employers names without industry, using EXCEPT
# then, counting the number of employers in the resulting table
query_5_4 = f'''
WITH
    employers_without_industry AS (
        SELECT id FROM employers
        EXCEPT
        SELECT employer_id FROM employers_industries
    )

SELECT COUNT(*) employers_without_industry_cnt FROM employers_without_industry
'''


In [32]:
# результат запроса
res = pd.read_sql_query(query_5_4, connection)
res

/var/folders/pz/qxrk46657v91nfg00jfysdx40000gn/T/ipykernel_13414/866972106.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  res = pd.read_sql_query(query_5_4, connection)


employers_without_industry_cnt
0                            8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [171]:
# текст запроса
query_5_5 = f'''
SELECT a.name AS employer
FROM employers a
JOIN employers_industries b
ON a.id = b.employer_id
GROUP BY 1
HAVING COUNT(b.employer_id) = 4
ORDER BY 1
OFFSET 2 -- Skip the first two companies
LIMIT 1 -- Retrieve only one record
'''

In [172]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

/var/folders/pz/qxrk46657v91nfg00jfysdx40000gn/T/ipykernel_13414/388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


employer
0     2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [35]:
# текст запроса
query_5_6 = f'''
SELECT COUNT(a.id) AS software_dev_employers_cnt
FROM employers a
JOIN employers_industries b
ON a.id = b.employer_id
JOIN industries c
ON b.industry_id = c.id
WHERE c.name = 'Разработка программного обеспечения'
'''

In [36]:
# результат запроса
res = pd.read_sql_query(query_5_6, connection)
res

/var/folders/pz/qxrk46657v91nfg00jfysdx40000gn/T/ipykernel_13414/615445897.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  res = pd.read_sql_query(query_5_6, connection)


software_dev_employers_cnt
0                        3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [37]:
# код для получения списка городов-милионников
import requests
from bs4 import BeautifulSoup

# Get the page by URL and parse its HTML
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
response = requests.get(url)
page = BeautifulSoup(response.text, 'html.parser')

# Find the table containing the city names
table = page.find('tbody')
# Final cities list
cities = []
# Skipping the header and iterating over the table rows
for row in table.find_all('tr')[1:]:
    city_name = row.find_all('td')[1].text
    # Adding each new city in the cities list
    cities.append(city_name)


In [173]:
# текст запроса
query_5_7 = f'''
(SELECT c.name AS region, COUNT(b.id) AS vacancies_cnt
FROM employers a
JOIN vacancies b
ON a.id = b.employer_id
JOIN areas c
ON b.area_id = c.id
WHERE a.name = 'Яндекс' -- filtering the company we're interested in
AND c.name IN ({' ,'.join(f"'{city}'" for city in cities)}) -- passing the list of parsed cities
GROUP BY 1 -- grouping the results by region
ORDER BY 2)

UNION ALL -- adding the total number of vacancies by the company across all regions

(SELECT 'Total', COUNT(b.id)
FROM employers a
JOIN vacancies b
ON a.id = b.employer_id
JOIN areas c
ON b.area_id = c.id
WHERE a.name = 'Яндекс'
AND c.name IN ({' ,'.join(f"'{city}'" for city in cities)}))
'''

In [174]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

/var/folders/pz/qxrk46657v91nfg00jfysdx40000gn/T/ipykernel_13414/1965596796.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


region  vacancies_cnt
0             Total            485
1              Омск             21
2         Челябинск             22
3        Красноярск             23
4         Волгоград             24
5             Пермь             25
6            Казань             25
7    Ростов-на-Дону             25
8               Уфа             26
9            Самара             26
10        Краснодар             30
11          Воронеж             32
12      Новосибирск             35
13  Нижний Новгород             36
14     Екатеринбург             39
15  Санкт-Петербург             42
16           Москва             54

***

### Выводы
- 4 из ТОП-5 компаний по количеству вакансий являются IT компаниями. И хоть Газпром Нефть не является IT компанией, мы можем отнести ее к диджитализированным компаниям, так как множество ее процессов завязаны на програмном обеспечении. **Гипотеза** - большинство вакансий нацелены на разработчиков, дизайнеров или менеджеров (project / product) для усиления IT департаментов компаний;

- Множество работодателей в качестве места расположения выбирают регионы, нежели конкретные города. В то же время, они размещают вакансии в конкретных городах. Необходимо уточнить, являются ли города в вакансиях фактическими рабочими локациями, или же работодатели могут предложить (настоять) другое место работы в пределах их региона. В последнем случае ожидания наших соискателей могут отличаться от ожиданий работодателей;

-  Яндекс является самой распределенной компанией с вакансиями в 181 городах и регионах. Всего компания опубликовала 485 вакансий в городах-миллионниках. Мы сможем предлагать вакансии Яндекса топовым разработчикам из нашего списка соискателей, даже если они живут на периферии и неготовы к релокации в отдаленное место. При условии, что в регионах Яндекс ищут именно разработчиков;

- Само по себе количество сфер деятельности, указанных у компании в анкете, ничего не говорит. Однако нам необходимо обратить внимание на те 8419 компаний, у которых в этом поле не указано ничего. Нам будет необходимо запонить эти пробелы, возможно, по ключевым словам их вакансий таких компаний;

- Наконец, в нашем списке имеется 3553 компании, занимающихся разработкой ПО. Вместе с другими IT компаниями, которые мы найдем после заполнения пробелов выше, мы сможем взять эти компании в subset, из которого мы будем рекомендовать подходящии вакансии в разработке для наших соискателей.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [40]:
# текст запроса
query_6_1 = f'''
SELECT COUNT(id) AS total_data_vacancies
FROM vacancies
WHERE LOWER(name) LIKE '%data%' OR LOWER(name) LIKE '%данн%'
'''

In [41]:
# результат запроса
res = pd.read_sql_query(query_6_1, connection)
res

/var/folders/pz/qxrk46657v91nfg00jfysdx40000gn/T/ipykernel_13414/1447873148.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  res = pd.read_sql_query(query_6_1, connection)


total_data_vacancies
0                  1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [52]:
# creating masks to use when filtering the query results with WHERE clause
# All DS roles
data_science_roles = ['data scientist', 'data science', 'исследователь данных', 'machine learning', 'машинн%обучен%']
# DS roles mask
data_science_mask = ' OR '.join([f"LOWER(name) LIKE '%{role}%'" for role in data_science_roles])
# ML mask, improtant to filter out HTML matches
ml_mask = "LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%'"
# junion mask
junior_mask = "LOWER(name) LIKE '%junior%' OR LOWER(experience) = 'нет опыта' OR LOWER(employment) = 'стажировка'"

# текст запроса
query_6_2 = f'''
SELECT COUNT(id) AS jr_ds_vacancies_cnt
FROM vacancies
WHERE (({data_science_mask}) OR ({ml_mask}))
AND ({junior_mask})
'''


In [53]:
# результат запроса
res = pd.read_sql_query(query_6_2, connection)
res

/var/folders/pz/qxrk46657v91nfg00jfysdx40000gn/T/ipykernel_13414/253039551.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  res = pd.read_sql_query(query_6_2, connection)


jr_ds_vacancies_cnt
0                   51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [58]:
# текст запроса
query_6_3 = f'''
SELECT COUNT(id) AS db_skills_required_cnt
FROM vacancies
WHERE (({data_science_mask}) OR ({ml_mask}))
AND (LOWER(key_skills) LIKE '%sql%' OR LOWER(key_skills) LIKE '%postgres%')
'''

In [59]:
# результат запроса
res = pd.read_sql_query(query_6_3, connection)
res

/var/folders/pz/qxrk46657v91nfg00jfysdx40000gn/T/ipykernel_13414/2273298653.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  res = pd.read_sql_query(query_6_3, connection)


db_skills_required_cnt
0                     229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [60]:
# текст запроса
# текст запроса
query_6_4 = f'''
SELECT COUNT(id) AS python_skills_required_cnt
FROM vacancies
WHERE (({data_science_mask}) OR ({ml_mask}))
AND LOWER(key_skills) LIKE '%python%'
''' 

In [61]:
# результат запроса
res = pd.read_sql_query(query_6_4, connection)
res

/var/folders/pz/qxrk46657v91nfg00jfysdx40000gn/T/ipykernel_13414/2140394540.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  res = pd.read_sql_query(query_6_4, connection)


python_skills_required_cnt
0                         357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [175]:
# текст запроса
query_6_5 = f'''
WITH -- CTE for creating a table for the number of skills required per a DS-related vacancy (nulls excluded)
    skills_per_vacancy AS (
        SELECT LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), '')) + 1 AS skills_cnt
        FROM vacancies
        WHERE (({data_science_mask}) OR ({ml_mask}))
        AND key_skills IS NOT NULL
    )

SELECT ROUND(AVG(skills_cnt), 2) AS avg_skills_required
FROM skills_per_vacancy
'''

In [176]:
# результат запроса
res = pd.read_sql_query(query_6_5, connection)
res

/var/folders/pz/qxrk46657v91nfg00jfysdx40000gn/T/ipykernel_13414/3031742892.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  res = pd.read_sql_query(query_6_5, connection)


avg_skills_required
0                 6.55

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [177]:
# текст запроса
query_6_6 = f'''
WITH -- CTE for calculating the DS salary per vacancy
    ds_salaries AS (
        SELECT
            experience,
        CASE -- cases for checking which part of the salary fork is present and calculating the salary for the vacancy accordingly
            WHEN salary_to IS NOT NULL AND salary_from IS NOT NULL THEN (salary_from + salary_to) / 2
            WHEN salary_to IS NULL AND salary_from IS NOT NULL THEN salary_from
            WHEN salary_to IS NOT NULL AND salary_from IS NULL THEN salary_to
        END AS salary
        FROM vacancies
        WHERE (salary_to IS NOT NULL OR salary_from IS NOT NULL)
        AND (({data_science_mask}) OR ({ml_mask}))   
    )

SELECT experience, ROUND(AVG(salary)) AS avg_salary
FROM ds_salaries
GROUP BY experience
ORDER BY 2 DESC
'''

In [178]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

/var/folders/pz/qxrk46657v91nfg00jfysdx40000gn/T/ipykernel_13414/519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience  avg_salary
0       От 3 до 6 лет    256454.0
1         Более 6 лет    157933.0
2  От 1 года до 3 лет    144233.0
3           Нет опыта     74643.0

***

### Выводы
- Из 1771 DS вакансий лишь 51 (2.88%) рассчитаны на junior level. Опять же, это может быть связано с тем, что работодатели ищут более проверенных специалистов. Возможно, нам лучше строить стратегию поиска работы для джунов через стажировки и личные договоренности с компаниями, нежели через платформы типа HH;

- Из 1771 DS вакансий лишь 229 (12.93%) требуют SQL или Postgres и лишь 357 (20.16%) требуют Python. **Гипотеза** - вакансии составлялись людьми, которые не особо знакомы с tech стаком дата сайнтиста. Возможно, эти люди работают на ролях нетехнического HR в компаниях-работодателях, и мы можем взять их экспертностью в вопросе поиске технических специалистов:
    1. Продать нашу экспертность
    2. Продать наш пул специалистов как высококачественный
    3. Начать продавать специалистов из пула
    4. Договориться с компаниями, что при открытии новых вакансий в DS они могут идти напрямую к нам. Мы им - качество и скорость, они нам - небольшой % c salary специалиста.
- Моя гипотеза выше подкрепляется тем, что несмотря на отсутствие требований к знанию SQL, Postgres и Python для дата сайнтистов у большинства компаний, они постят ~6.55 требуемых навыков. Нужно изучать этот вопрос глубже: возможно они там пишут общие или нерелевантные требования, а возможно у них просто другой tech стак, что заставит нас пересмотреть наши требования к tech стаку соискателей;
    - **ВНИМАНИЕ!!!** Платформа говорит, что мой результат неверен. Пожалуйста, приложите ссылку на правильный запрос, чтобы я понял, где ошибка. Перепроверял много раз, пытался по-другому запрос сформировать - не помогло.

- Средние вилки зарплат дата сайнтистов на HH выглядят логично, за исключением низкой зарплаты для специалистов с опытом работы 6+ лет (150K рублей для синьора - несерьезно). Для джунов и мидлов мы можем давать реалистичные прогнозы нашим соискателям, фильтруя или "выравнивая" завышенные или заниженные ожидания. Также мы можем сравнить HH как канал продажи наших специалистов с альтернативами для более оптимального распределения ресурсов.
    - **ВНИМАНИЕ!!!** Платформа говорит, что мой результат неверен. Пожалуйста, приложите ссылку на правильный запрос, чтобы я понял, где ошибка. Перепроверял много раз, пытался по-другому запрос сформировать - не помогло.

# Общий вывод по проекту

### Общие выводы
1. В основном, работодатели на HH ищут специалистов уровня junior-middle. Для ролей в DS зарпалатная вилка для джунов и недавно состоявшихся мидлвов (от "нет опыта" до 3 лет) - 109,438 рублей, а для "крепких мидлов" и недавно состоявшихся синьоров (от 3 до 6 лет) - 256,454 рубля.
    - Эти диапазоны позволяют нам провести начальный отбор потенциально подходящих специалистов из нашего пула, синхронизируя их опыт и зарплатные ожидания с рынком.

2. В ТОП-5 работодателей по количеству вакансий входят 4 IT компании (Яндекс, Ростелеком, Тинькофф и СБЕР) и 1 диджитализированная компания (Газпром Нефть). При условии, что эти компании ищут разработчиков и дата сайнтистов в частности, мы сможем предлагать нашим соискателям работу в престижных организациях с очевидной возможностью карьерного роста. Это может стать одним из наших конкурентных преимуществ: если наши кандидаты успешно пройдут стадии отбора и покажут сильный старт в этих компаниях, то мы получим возможность стать одним из outsourced HR отделов этих крупных компаний.

3. Необходимо провести дополнительное исследование, чтобы понять требования к основным навыкам дата сайнтистов в престижных организациях. Это поможет нам понять, насколько HR тех компаний понимают требуемый tech стак и свериться с нашими требованиями к соискателям. Напомню, что из 1771 вакансий в DS лишь у 12.93% были требования к знанию SQL или Postgres и лишь у 20.16% к знанию Python.
    - Если выясниться, что вакансии составляются, публикуются и курируются нетехническим HR'ом, то нам будет проще продать нашу экспертность через помощь составления "правильных вакансий" и предложения подходящий кандидатов из нашего пула. Оставлю нюансы нашим продажникам.

4. Вообще, нам имеет смысл как можно скорее установить доверительные отношения с Яндексом, так как он лидирует по уровню присутствия в регионах и городах (вакансии в 181 местах), с 20+ вакансиями в каждом их городов-миллионников России.
    - Для Яндекса: поставка высококвалифицированных специалистов от нас;
    - Для нас: возможность привлекать больше амбициозных соискателей и зарабатывать больше с % от контрактов;
    - Для соискателей: возможность попасть в крупную IT компанию через наши "теплые контакты" с HR Яндекса.
***
### Варианты продолжения исследования
- Заполнить нулевое поле индустрии у 8419 компаний, чтобы понять, какие из них работают в сфере IT, тем самым расширив пул подходящих для наших соискателей компаний;

- Понять, кого именно в каких городах ищет Яндекс и сформировать стратегию и тактику по тому, как и с какими кандидатами заходить к его HR'у;

- Понять, почему у большинства DS вакансий отсутствуют требования к знанию SQL/Postgres/Python - неопытность HR или же специфический tech stack;

- Соединить базу работодателей и вакансий с базой соискателей для синхронизации спроса рынка с нашим предложением. Все это потребуется сравнить с соответствующим распределением кандидатов в нашем пуле:
    - Каких специалистов и на какие роли ищет рынок;
    - Какой требуемый опыт работы и требования к hard skills / teck stack;
    - Какой тип занятости (FT / PT / project-base) и требования к локации (on-site / remote / hybrid);
    - Чем больше мэтч с запросами рынка, тем быстрее мы можем реагировать на открытые вакансии и тем больше мы можем зарабатывать.
